"Hey Luna" Wakeword with more advanced notebook settings

dreimer1986 opened the pull request:
Created with the help of https://github.com/MasterPhooey/MicroWakeWord-Trainer-Docker under Windows 11 Docker and 16h of waiting... and thus obviously with way more demanding values. The result is quite fine IMO. Thus I ask if you wanna use it

dreimer1986 committed:
Add Luna Wake Word Test Version

dreimer1986 committed:
Add Luna as v2

dreimer1986 committed:
Naming

Proper naming of files and wakeword itself.

Marcel0024 commented on models/v2/hey_luna.json:
```diff
@@ -0,0 +1,18 @@
+  ],
+  "version": 2,
+  "micro": {
+    "probability_cutoff": 0.97,
```
is 0.97 a bit high?

andreasbrett commented:
I have yet to test it but this should be great for non-native English speakers as dialectic variances should be considerably low for "Luna" (and of course "hey").

dreimer1986 commented:
I agree. I use it all the time here. Even added the new needed parameters based on the average of the three official ones. No clue how to get the real ones, but I did not wanna remove my wake word after the recent updates, too. Thus I tried to get some that most likely do not cause big harm and succeeded. Of course it reacts to stuff not meant for it, but my Alexa is even worse with that. So I for myself am happy with the results.

morbitangel commented:
I've used this for 2 days now and i havent had a 'false' wake yet pretty impressive.
Would you mind sharing the settings you used in the advanced notebook to train this model?

dreimer1986 commented:
I had them somewhere... Gonna look for them and then post them here

EDIT:
Could not find everything yet I did, but I took inspiration from the workbook for openWakeWord generation here.
First the phonetic transcribe used: hey_loo_nuh

And now the tricky ones:
--max-samples and --batch-size and a few steps later the line config["training_steps"] = [10000] ... I am 99.9% sure that I chose 35000 as max samples. But the batch one is completely gone off my mind... I think I followed the values from here https://colab.research.google.com/drive/1q1oe2zOyZp7UsB3jJiQ1IFn8z5YfjwEb?usp=sharing#scrollTo=1cbqBebHXjFD and thus the 35000, but did I change the batch size? That I completely forgot. Here what comes to my mind immediately when I look at it:

100 as batch size stays and thus we get 300 batches of samples. Way down below is the line
config["training_steps"] = [10000] and this one I updated to 15000 (at least I am quite sure of that).

Btw, it seems like the openWakeWord workbook which I used way before the MicroWakeWord one to create a WakeWord seems to create tflite files, too. Looks like I have to do some more revisions of this Wake Word soon. ^

dreimer1986 commented:
OK, I decided to create a new wake word file with documented data this time. But tbh, the documentation just says that I hope that using the unmodified advanced notebook that is said to need 20 GB VRAM in a monster GFX card does not take weeks to finish on my CPU. I prepped everything and right now it finished all the creation and is now working on training the model itself. I will post here as soon as it's done. But as little comparison... The advanced notebook uses 50000 created samples and 40000 training steps and more advanced negative weight for a less forgiving model. So if it finishes in a acceptable time frame I would have a way better trained wake word than my 35000/15000 one from before.

First I tried to get it running on my Steam Deck which rather became my main PC these days... But this is not really working out ^^

2025-07-12 22:10:00.963195: W external/local_xla/xla/tsl/framework/cpu_allocator_impl.cc:83] Allocation of 1424000000 exceeds 10% of free system memory.
Killed

morbitangel commented:
yeh i was confused when i read your settings since the advanced notebook ive been using has those settings already higher.
it doesnt need 20GB VRAM though ive ran it on my 4060TI with 16GB VRAM just fine.
ill take a look at running 'hey luna' through the advanced notebook today it shouldnt take superlong with a GPU last time i did with 75000 training samples and 40000 training steps it took about two hours i think to finish training a wake word.

edit:

 currently training your wake word with the advanced notebook on its default settings aka 50000 samples and 40000 training steps.
i'll post it here when its done

since i cant attach tflite files here is the link to my repo for it ill leave it public
https://raw.githubusercontent.com/morbitangel/friendly-doodle/refs/heads/main/hey_luna.json

let me know how well it works if needed im willing to train it with more samples/steps

dreimer1986 commented:
I am at 19500 of 40000 now. Should be done today evening. Just as comparison. It took a day on my 16 core Ryzen then. ^^ Looking forward to see if the result is identical or at least similar.

dreimer1986 committed:
Second try for "Hey Luna"

Made with the advanced notebook from https://github.com/MasterPhooey/MicroWakeWord-Trainer-Docker.
Compared to the former experiment where I hand edited a few parameters for more samples and more runs (35000/15000) this one uses the default values that are already set up way higher (50000/40000) and the failure penalty is set higher, too. Result is being tested now.
target_word = 'hey_loo_nuh'
Took about 16 hours on CPU only as the notebook does not allow GPU optimization on my Intel Arc GPU.

dreimer1986 commented:
Looking forward to your results. My second try seems to react a bit too often to my environment with Anime in Japanese Dub here. But I am still tinkering with the probability cutoff value that was added to Voice PE. This might be the way to get it running fine.

But yeah, we need a few testers who test your new one, my v2 one and best, tinker with the cutoff values a bit. I don't really care who got the better results in the end. I just want to have my wake word working fine for not just me, but all out there and that one to be added here.

Ah, for the interested ones, this is how I add my model to the firmware right now. Before that I chose cutoff values ~15 lower than the ones right now. I just took the middle between the lowest and the highest values already in there. Now I went to the model with the highest values to tinker a bit. Just by increasing the values a bit it started to behave way better than before.

```
substitutions:
  name: home-assistant-voice-09097c
  friendly_name: Home Assistant Voice 09097c
packages:
  Nabu Casa.Home Assistant Voice PE: github://esphome/home-assistant-voice-pe/home-assistant-voice.yaml
esphome:
  name: ${name}
  name_add_mac_suffix: false
  friendly_name: ${friendly_name}
api:
  encryption:
    key: 30W/pJym8+1MDCDP3aGEQxA5TMr5W5x1HPeRBiup5k0=

wifi:
  ssid: !secret wifi_ssid
  password: !secret wifi_password

micro_wake_word:
  models:
    - model: https://raw.githubusercontent.com/dreimer1986/micro-wake-word-models/refs/heads/main/models/v2/hey_luna.json
      id: hey_luna
    - model: https://raw.githubusercontent.com/dreimer1986/micro-wake-word-models/refs/heads/main/models/v2/hey_luna_v2.json
      id: hey_luna_v2

select:
  - platform: template
    name: "Wake word sensitivity"
    optimistic: true
    initial_option: Moderately sensitive
    restore_value: true
    entity_category: config
    options:
      - Slightly sensitive
      - Moderately sensitive
      - Very sensitive
    on_value:
      lambda: |-
        if (x == "Slightly sensitive") {
          id(okay_nabu).set_probability_cutoff(217);    // 0.85 -> 0.000 FAPH on DipCo (Manifest's default)
          id(hey_jarvis).set_probability_cutoff(247);   // 0.97 -> 0.563 FAPH on DipCo (Manifest's default)
          id(hey_mycroft).set_probability_cutoff(253);  // 0.99 -> 0.567 FAPH on DipCo
          id(hey_luna).set_probability_cutoff(253);
          id(hey_luna_v2).set_probability_cutoff(253);
        } else if (x == "Moderately sensitive") {
          id(okay_nabu).set_probability_cutoff(176);    // 0.69 -> 0.376 FAPH on DipCo
          id(hey_jarvis).set_probability_cutoff(235);   // 0.92 -> 0.939 FAPH on DipCo
          id(hey_mycroft).set_probability_cutoff(242);  // 0.95 -> 1.502 FAPH on DipCo (Manifest's default)
          id(hey_luna).set_probability_cutoff(242);
          id(hey_luna_v2).set_probability_cutoff(242);
        } else if (x == "Very sensitive") {
          id(okay_nabu).set_probability_cutoff(143);    // 0.56 -> 0.751 FAPH on DipCo
          id(hey_jarvis).set_probability_cutoff(212);   // 0.83 -> 1.502 FAPH on DipCo
          id(hey_mycroft).set_probability_cutoff(237);  // 0.93 -> 1.878 FAPH on DipCo
          id(hey_luna).set_probability_cutoff(237);
          id(hey_luna_v2).set_probability_cutoff(237);
        }
```

morbitangel commented:
Mine also seemed to wake more from the TV shows i have running pretty much 24/7.
I've been using 2 S3-Box-3's though and not a VPE so i dont have the wake word sensitivity config yet.
I just dusted off my VPE and installed it to see how it will perform (i dont use the VPE much cos of the weird accoustics here messing with STT too much)

dreimer1986 committed:
Tried to slow the ww down a bit...

... hey_____loo__nuh is the test now

morbitangel commented:
Ported over the wake word sensitivity config to my S3 boxes with some help from others.
Seems to work pretty well so far only very rarely wakes from my TV now maybe once a day.
i do have to speak a bit louder to get it to recognize me but i can live with that trade off.

dreimer1986 committed:
Merge branch 'esphome:main' into main

sahilph commented:
Thanks @dreimer1986  This works perfectly.  Hopefully it gets merged soon

janstadt commented:
ive started noticing hey luna in a few other device implementations of wakewords. Would love to see this get mainlined so we can pull it in to box3s without workarounds. any status updates?
